addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionManager.java

```diff
@@ -761,430 +561,322 @@ public PathsPromoteResult rollbackPathsPromote( final PathsPromoteResult result
             }
             catch ( InterruptedException | ExecutionException e )
             {
-                logger.error( "Path promotion rollback failed: " + request.getSource() + " -> " + request.getTargetKey(), e );
-                throw new PromotionException( "Execution of path promotion rollback failed.", e );
+                logger.error( "Path promotion rollback failed. From (target): " + request.getTarget()
+                                              + ", to (source): " + request.getSource(), e );
+                throw new PromotionException( "Path promotion rollback failed.", e );
             }
         }
     }
 
     private Future<PathsPromoteResult> submitRollbackPathsPromote( PathsPromoteResult result )
-            throws IndyWorkflowException
+                    throws IndyWorkflowException
     {
         return detectOverload( () -> asyncPromotionService.submit( () -> {
+            if ( result.getCompletedPaths().isEmpty() )
+            {
+                return result;
+            }
+
+            /*
+             * Rollback is the opposite of promotion. We just revert the src and target, set the paths
+             * to the completed, and call the doPathsPromotion.
+             */
+            PathsPromoteRequest request = result.getRequest();
+            PathsPromoteRequest newRequest = new PathsPromoteRequest( request.getTarget(), request.getSource(),
```

Comment:
Exactly. For requests with purge = false there should be deletions performed instead of copies. So swapping source and target won't do anything there.

Reply:
I haven't checked that part, but I would expect what is skipped does not get purged. If it's not the case then ok, but should be documented, because the behaviour is not clear at this point.

Reply:
The rollback only purges the prev "completed" paths (those were files freshly copied in prev promotion). It does not touch or change the skipped ones.